app/controllers/api/pxe_servers_controller.rb

```diff
@@ -48,9 +48,18 @@ def edit_resource(type, id, data)
       menus = data.delete('pxe_menus')
       authentication = data.delete('authentication')
       PxeServer.transaction do
-        if menus
+        # When server URI changes all menus have to be discarded, sorry.
+        # But when URI remains we can perserve menus, as long as they aren't changed too.
+        if data['uri'] != server.uri
           server.pxe_menus.destroy_all
-          data['pxe_menus'] = create_pxe_menus(menus)
+          data['pxe_menus'] = create_pxe_menus(menus || [])
+        elsif menus
+          desired_menus = menus.map { |el| el['file_name'] }
+          current_menus = server.pxe_menus.map(&:file_name)
+          to_keep, to_destroy = current_menus.partition { |menu| desired_menus.include?(menu) }
+          to_create = desired_menus.reject { |menu| current_menus.include?(menu) }
+          server.pxe_menus.where(:file_name => to_destroy).destroy_all
+          data['pxe_menus'] = server.pxe_menus.where(:file_name => to_keep) + create_pxe_menus(to_create.map { |f| { 'file_name' => f } })
```

Comment:
feels a bit heavy on business logic here, I wonder if this should not be provided by a model helper method that we'd leverage.

Reply:
I think you're right @abellotti, here'se the utility method https://github.com/ManageIQ/manageiq/pull/19134